### PR TITLE
Add more E2E tests for external connectivity use case

### LINKED ISSUE #1581: Add more E2E tests for external connectivity use case
**What would you like to be added**:
More E2E tests for external connectivity use case.

**Why is this needed**:
https://github.com/submariner-io/submariner/pull/1202 and https://github.com/submariner-io/submariner/pull/1455 have  added initial tests for this use case, however, more tests are needed to make the coverage of the tests similar to cross-cluster connectivity.

### PR DESCRIPTION
**Pull request by mkimuram**
This PR adds more e2e tests for external connectivity use case. Tests are refactored to align with E2E tests for normal cross-cluster use cases, before adding tests.

Note that most codes are copied from https://github.com/submariner-io/submariner/pull/1558 , so that #1558 can focus on adding a new feature.

Dependency:
- [x] https://github.com/submariner-io/submariner/pull/1695

Fixes: #1581

### CODE REVIEW

File: test/external/dataplane/connectivity.go [hunk 1]
```diff
@@ -40,165 +40,179 @@ var (
...
-		By(fmt.Sprintf("Sending an http request from the test pod %q %q in cluster %q to the external app's ingressGlobalIP %q",
-			np.Pod.Name, podGlobalIP, clusterName, extIngressGlobalIP))
+	By("Verifying the pod received the request")
 
-		cmd := []string{"curl", "-m", "10", fmt.Sprintf("%s:%d/%s%s", extIngressGlobalIP, 80, f.Namespace, clusterName)}
-		_, _ = np.RunCommand(cmd)
+	podLog := np.GetLog()
 
-		By(fmt.Sprintf("Verifying that external app received request from egressGlobalIP %q", podGlobalIP))
+	if clusterName == externalClusterName {
```
Comment (sridhargaddam) [inline test/external/dataplane/connectivity.go]: Can you please add some comments on why the validation is different if the clusterName matches with externalClusterName.

File: test/external/dataplane/connectivity.go [hunk 2]
```diff
@@ -40,165 +40,179 @@ var (
...
+				Networking:        networking,
+				Cluster:           cluster,
+				ClusterScheduling: clusterScheduling,
+			})
+		})
+	}
 
-func testExternalConnectivity(f *framework.Framework) {
-	externalClusterName := getExternalClusterName(framework.TestContext.ClusterIDs)
+	When("In an external-app-connected cluster", func() {
```
Comment (sridhargaddam) [inline test/external/dataplane/connectivity.go]: When the test case is executed, AFAIU the sentence will not be correct as you are using When and starting the sentence with "In ...". 

@tpantelis can you please suggest?

File: test/external/dataplane/connectivity.go [hunk 3]
```diff
@@ -228,3 +242,16 @@ func getExternalClusterIndex(names []string) framework.ClusterIndex {
 	// TODO: consider right error handling
 	return framework.ClusterIndex(0)
```
Comment (sridhargaddam) [inline test/external/dataplane/connectivity.go]: Instead of returning framework.ClusterIndex(0) which could be wrong and misleading when the test-case fails. I think it might be better to fail the test-case. What do you say?

File: test/external/dataplane/connectivity.go [hunk 4]
```diff
@@ -228,3 +242,16 @@ func getExternalClusterIndex(names []string) framework.ClusterIndex {
...
+func getNonExternalClusterIndex(names []string) framework.ClusterIndex {
+	clusterName := getExternalClusterName(names)
+
+	for idx, cid := range names {
+		if cid != clusterName {
+			return framework.ClusterIndex(idx)
+		}
+	}
+
+	// TODO: consider right error handling
```
Comment (sridhargaddam) [inline test/external/dataplane/connectivity.go]: Same as above

File: test/external/dataplane/connectivity.go [hunk 5]
```diff
@@ -40,165 +40,179 @@ var (
...
+			})
+		})
 
-		command := []string{"curl", "-m", "10", fmt.Sprintf("%s:%d/%s%s", svcIP, 80, f.Namespace, clusterName)}
-		_, _ = docker.RunCommand(command...)
+		// Access from a hostnetwork pod to external apps is not supported
+	})
 
-		By("Verifying the pod received the request")
+	When("In a non-external-app-connected cluster", func() {
```
Comment (sridhargaddam) [inline test/external/dataplane/connectivity.go]: Its not clear what non-external-app-connected cluster implies.

File: test/external/dataplane/gn_connectivity.go [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+		})
+
+		When("a pod with HostNetworking connects via TCP to the globalIP of a remote service", func() {
+			BeforeEach(func() {
+				toEndpointType = tcp.GlobalServiceIP
+				networking = framework.HostNetworking
+				egressIPType = subFramework.ClusterSelector
+			})
+
+			// Access from a pod with HostNetworking fails if a pod is not on NonGateway nodes
```
Comment (sridhargaddam) [inline test/external/dataplane/gn_connectivity.go]: Can you explain what is the problem with that use-case?

**Review by sridhargaddam — CHANGES_REQUESTED** [review]
Thanks for the PR @mkimuram.
I think we have to capture the topology of the external network setup or improve the comments/description of the tests, which otherwise will be very hard to follow for anyone looking at the code.

**Comment by sridhargaddam** [comment]
@mkimuram Also, please fix the conflicts.

File: test/external/dataplane/connectivity.go (hunk 1 above)
Comment (mkimuram) [reply test/external/dataplane/connectivity.go]: Added an ascii art of the test environment and expected behaviors.

File: test/external/dataplane/connectivity.go (hunk 2 above)
Comment (mkimuram) [reply test/external/dataplane/connectivity.go]: Changed to `When("connected from a gateway-cluster"`.  The word is the same to the ascii art. Any ideas to improve the wording are welcomed.

File: test/external/dataplane/connectivity.go (hunk 3 above)
Comment (mkimuram) [reply test/external/dataplane/connectivity.go]: Modified to return error as the second return value and handle the error in caller.

File: test/external/dataplane/gn_connectivity.go (hunk 6 above)
Comment (mkimuram) [reply test/external/dataplane/gn_connectivity.go]: It just fails to connect and I'm still not sure the reason. If we have the use case for these configurations, we need to investigate more and consider improvement.  I would like to make it just documented and leave as is.

**Comment by mkimuram** [comment]
@sridhargaddam 

Rebased and addressed comments.

File: test/external/dataplane/gn_connectivity.go [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+    kind                            *-----+--------+-------*
+
+
+   - For globalnet environment, expected behaviors of connectivity and source IPs are:
+
+       from / to       |    ext-app    |  gateway-cluster  |  non-gateway-cluster  |
+   ------------------- | ------------- | ----------------- | --------------------- |
+   ext-app             |    N/A        |       R(*2)(*4)   |        S(*3)          |
+   gateway cluster     |    R(*2)(*4)  |       N/A         |        S(*1)          |
+   non gateay cluster  |    S(*3)      |       S(*1)       |        N/A            |
```
Comment (nerdalert) [inline test/external/dataplane/gn_connectivity.go]: quick typo `gateay`

Comment (mkimuram) [reply test/external/dataplane/gn_connectivity.go]: Fixed and squashed the commit. Thank you for pointing it out.

File: test/external/dataplane/connectivity.go [hunk 8]
```diff
@@ -38,179 +38,216 @@ const (
...
-			np.Pod.Name, podIP, clusterName, dockerIP))
+		When("a pod connects via TCP to a remote pod", func() {
+			BeforeEach(func() {
+				toEndpointType = tcp.PodIP
+				networking = framework.PodNetworking
+			})
 
-		cmd := []string{"curl", "-m", "10", fmt.Sprintf("%s:%d/%s%s", dockerIP, 80, f.Namespace, clusterName)}
-		_, _ = np.RunCommand(cmd)
+			// Access from a pod on NonGatewayNodes to external apps is not supported for an external-app-connected cluster
```
Comment (sridhargaddam) [inline test/external/dataplane/connectivity.go]: As discussed during the meeting, can you please explain why access from pods on nonGW nodes to ext-app does not work?

File: test/external/dataplane/connectivity.go [hunk 9]
```diff
@@ -38,179 +38,216 @@ const (
...
-		docker := framework.New(extAppName)
+   Legend: N: Not reachable, R: Reachable (source IP isn't globalIP),
+           S: Source IP is global IP (and reachable)
 
-		// Get IPs to use later.
-		podIP := np.Pod.Status.PodIP
-		svcIP := svc.Spec.ClusterIP
-		dockerIP := docker.GetIP(extNetName)
+   (*1) Not covered in this test, but covered in normal connectivity tests.
+   (*2) Pod w/ hostnetwork isn't reachable.
```
Comment (sridhargaddam) [inline test/external/dataplane/connectivity.go:71]: When you say "Pod w/ hostnetwork isn't reachable", is the hostnetwork Pod you are talking the sourcePod (aka client) or the destinationPod (aka Server)?

File: test/external/dataplane/connectivity.go [hunk 10]
```diff
@@ -38,179 +38,216 @@ const (
...
-		remoteIP := f.AwaitGlobalIngressIP(np.Config.Cluster, svc.Name, svc.Namespace)
-		Expect(remoteIP).ToNot(Equal(""))
+		// Access from a hostnetwork pod to external apps is not supported
+	})
+})
 
-		podGlobalIPs := f.AwaitClusterGlobalEgressIPs(np.Config.Cluster, constants.ClusterGlobalEgressIPName)
-		Expect(podGlobalIPs).ToNot(BeEmpty())
+func testExternalConnectivity(p testParams) {
+	externalClusterName := getGatewayClusterName(framework.TestContext.ClusterIDs)
```
Comment (sridhargaddam) [inline test/external/dataplane/connectivity.go]:
```suggestion
	gateway-cluster := getGatewayClusterName(framework.TestContext.ClusterIDs)
```

To match with the topology, lets use gateway-cluster instead of externalCluster.

File: test/external/dataplane/gn_connectivity.go [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+   gateway-cluster     |    R(*2)(*4)  |       N/A         |        S(*1)          |
+   non-gateway-cluster |    S(*3)      |       S(*1)       |        N/A            |
+
+   Legend: N: Not reachable, R: Reachable (source IP isn't globalIP),
+           S: Source IP is global IP (and reachable)
+
+   (*1) Not covered in this test, but covered in normal connectivity tests.
+   (*2) Pod w/ hostnetwork isn't reachable.
+   (*3) Pod w/ hostnetwork isn't reachable when pod isn't on a gateway node.
+   (*4) Pod isn't reachable, when pod isn't on a gateway node
```
Comment (sridhargaddam) [inline test/external/dataplane/gn_connectivity.go:66]:
```suggestion
   (*4) Pod isn't reachable, when pod isn't on a gateway node -> Can you please explain why this use-case does not work?
```

File: test/external/dataplane/gn_connectivity.go [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+    pseudo-ext *-----+------------+--*    |        |
+                                          |        |
+    kind                            *-----+--------+-------*
+
+
+   - For globalnet environment, expected behaviors of connectivity and source IPs are:
+
+       from / to       |    ext-app    |  gateway-cluster  |  non-gateway-cluster  |
+   ------------------- | ------------- | ----------------- | --------------------- |
+   ext-app             |    N/A        |       R(*2)(*4)   |        S(*3)          |
```
Comment (sridhargaddam) [inline test/external/dataplane/gn_connectivity.go:56]: When you say S(*3) -> Pod w/ hostnetwork is not reachable when pod isn't on a gateway node -> Does it mean the ext-app is trying to connect to a Pod w/ hostnetwork?

If it is so, then such use-case may not work even between regular clusters connected via submariner because the hostnetwork pod will have an IPAddress from the underlay network which is not advertised as local subnet to the remote clusters. 

Please clarify.

File: test/external/dataplane/connectivity.go [hunk 13]
```diff
@@ -38,179 +38,216 @@ const (
...
-			testExternalConnectivity(f)
-		}
-	})
-})
+/*
+   - Test environment for external network w/o globalnet is as follows:
 
-func testExternalConnectivity(f *framework.Framework) {
-	externalClusterName := getExternalClusterName(framework.TestContext.ClusterIDs)
+                  [ext-app]  [gateway-cluster]  [non-gateway-cluster]
```
Comment (sridhargaddam) [inline test/external/dataplane/connectivity.go:52]: I'm not sure if I'm the only one who feels that "gateway-cluster" and "non-gateway-cluster" could create confusion with "gateway-node" and "non-gateway-node". @tpantelis @dfarrell07 do you have any suggestions?

Basically, the idea behind "gateway-cluster" is that its a cluster that is directly connected to the external-app.

File: test/external/dataplane/connectivity.go [hunk 14]
```diff
@@ -38,179 +38,216 @@ const (
...
+				toEndpointType = tcp.PodIP
+				networking = framework.PodNetworking
+			})
 
-		cmd := []string{"curl", "-m", "10", fmt.Sprintf("%s:%d/%s%s", dockerIP, 80, f.Namespace, clusterName)}
-		_, _ = np.RunCommand(cmd)
+			// Access from a pod on NonGatewayNodes to external apps is not supported for an external-app-connected cluster
 
-		By("Verifying that external app received request")
+			When("the pod is on a gateway", func() {
```
Comment (sridhargaddam) [inline test/external/dataplane/connectivity.go]: The output of the e2e will be as follows
```
[external-dataplane] Connectivity when connected from a gateway-cluster when a pod connects via TCP to a remote pod when the pod is on a gateway 
  should be able to connect from an external app to a pod in a cluster
...
```

Also, the test-case validates connectivity in **both** the directions - i.e., from the Pod which is running either on the gateway-cluster or non-gateway-cluster to the ext-app (aka docker). But the description does not convey that meaning.

Comment (sridhargaddam) [reply test/external/dataplane/connectivity.go]: Please use "Context" where necessary to frame an appropriate sentence for each of the test-case.

File: test/external/dataplane/connectivity.go (hunk 1 above)
Comment (sridhargaddam) [reply test/external/dataplane/connectivity.go]: Thank you

**Review by sridhargaddam — CHANGES_REQUESTED** [review]
Thanks @mkimuram. Mostly my comments are related to the test description and about the naming convention used in the topology.

**Comment by sridhargaddam** [comment]
@mkimuram When I'm trying to run these test-cases on my laptop, I'm getting the following errors (for all the external tests). Is it because of any docker version mismatch, any idea?

```
[external-dataplane] when an ext-app tries to connect to a Pod running on the gateway node of gateway-cluster when a pod connects via TCP to a remote service when the pod is on a gateway 
  should be able to connect from an external app to a pod in a cluster
  /go/src/github.com/submariner-io/submariner/test/external/dataplane/connectivity.go:83
STEP: Creating namespace objects with basename "ext-dataplane"
STEP: Generated namespace "e2e-tests-ext-dataplane-4s62f" in cluster "cluster2" to execute the tests in
STEP: Creating namespace "e2e-tests-ext-dataplane-4s62f" in cluster "cluster1"
STEP: Creating a pod and a service in cluster "cluster1"
STEP: Deleting namespace "e2e-tests-ext-dataplane-4s62f" on cluster "cluster2"
STEP: Deleting namespace "e2e-tests-ext-dataplane-4s62f" on cluster "cluster1"

• Failure [6.458 seconds]
[external-dataplane]
/go/src/github.com/submariner-io/submariner/test/external/dataplane/connectivity.go:74
  when an ext-app tries to connect to a Pod running on the gateway node of gateway-cluster
  /go/src/github.com/submariner-io/submariner/test/external/dataplane/connectivity.go:99
    when a pod connects via TCP to a remote service
    /go/src/github.com/submariner-io/submariner/test/external/dataplane/connectivity.go:118
      when the pod is on a gateway
      /go/src/github.com/submariner-io/submariner/test/external/dataplane/connectivity.go:126
        should be able to connect from an external app to a pod in a cluster [It]
        /go/src/github.com/submariner-io/submariner/test/external/dataplane/connectivity.go:83

        Unexpected error:
            <*exec.ExitError | 0xc00057e920>: {
                ProcessState: {
                    pid: 6695,
                    status: 256,
                    rusage: {
                        Utime: {Sec: 0, Usec: 312039},
                        Stime: {Sec: 0, Usec: 46765},
                        Maxrss: 68444,
                        Ixrss: 0,
                        Idrss: 0,
                        Isrss: 0,
                        Minflt: 15819,
                        Majflt: 0,
                        Nswap: 0,
                        Inblock: 0,
                        Oublock: 0,
                        Msgsnd: 0,
                        Msgrcv: 0,
                        Nsignals: 0,
                        Nvcsw: 251,
                        Nivcsw: 86,
                    },
                },
                Stderr: nil,
            }
            exit status 1
        occurred

        /go/src/github.com/submariner-io/submariner/vendor/github.com/submariner-io/shipyard/test/e2e/framework/docker.go:47

        Full Stack Trace
        github.com/submariner-io/shipyard/test/e2e/framework.(*Docker).GetIP(0xc00065f0f8, 0x162f87e, 0xa, 0x6, 0x50)
                /go/src/github.com/submariner-io/submariner/vendor/github.com/submariner-io/shipyard/test/e2e/framework/docker.go:47 +0x20d
        github.com/submariner-io/submariner/test/external/dataplane.testExternalConnectivity(0xc00035fef0, 0x1, 0x0, 0x1, 0x1)
                /go/src/github.com/submariner-io/submariner/test/external/dataplane/connectivity.go:199 +0x33e
        github.com/submariner-io/submariner/test/external/dataplane.glob..func1.1.1()
                /go/src/github.com/submariner-io/submariner/test/external/dataplane/connectivity.go:89 +0xd2
        github.com/submariner-io/shipyard/test/e2e.RunE2ETests(0xc000498180, 0x4bfc1adff06a)
                /go/src/github.com/submariner-io/submariner/vendor/github.com/submariner-io/shipyard/test/e2e/e2e.go:95 +0x115
        github.com/submariner-io/submariner/test/external_test.TestE2E(0xc000498180)
                /go/src/github.com/submariner-io/submariner/test/external/e2e_external_test.go:29 +0x2b
        testing.tRunner(0xc000498180, 0x16d4ec0)
                /usr/lib/golang/src/testing/testing.go:1193 +0xef
        created by testing.(*T).Run
                /usr/lib/golang/src/testing/testing.go:1238 +0x2b3

```

**Comment by sridhargaddam** [comment]
Hello @mkimuram, can you take a look at the review comments. Let us know if you need any additional info.

**Comment by mkimuram** [comment]
To run the tests in this PR, please type `make e2e using=external-net,globalnet`.

**Comment by dfarrell07** [comment]
> To run the tests in this PR, please type `make e2e using=external-net,globalnet`.

Should we add a GitHub Action to run that for us in CI, @mkimuram? I can add it for you if you'd like.

**Comment by dfarrell07** [comment]
> > To run the tests in this PR, please type `make e2e using=external-net,globalnet`.
> 
> Should we add a GitHub Action to run that for us in CI, @mkimuram? I can add it for you if you'd like.

Ah we have that already! Just needs the ready-to-test label to run.

**Comment by dfarrell07** [comment]
It seems that we're not overriding the default Ginkgo ignore rule for skipping tests tagged `external-dataplane`, however.

```
2022-02-03T18:06:48.5214901Z ##[group]Run submariner-io/shipyard/gh-actions/e2e@devel
2022-02-03T18:06:48.5215405Z with:
2022-02-03T18:06:48.5215676Z   using: external-net 
2022-02-03T18:06:48.5216014Z   k8s_version: 1.23
2022-02-03T18:06:48.5216307Z   working-directory: .
2022-02-03T18:06:48.5216635Z   target: e2e
2022-02-03T18:06:48.5216885Z   focus: \[.*\]
2022-02-03T18:06:48.5217370Z   skip: \[external-dataplane\]
2022-02-03T18:06:48.5217704Z   testdir: test/e2e
2022-02-03T18:06:48.5217955Z ##[endgroup]
```

https://pipelines.actions.githubusercontent.com/XY3uESDBHEDMVySv1XHhnjBBj2HTkVMErmGSbcdh0GWkFVF5nH/_apis/pipelines/1/runs/18224/signedlogcontent/11?urlExpires=2022-02-03T18%3A25%3A02.3704302Z&urlSigningMethod=HMACV1&urlSignature=ND68tdQGb%2BAAa5DX5jeWa4nEpnC79tOxLV%2FlW5mdZ%2Bc%3D

**Comment by dfarrell07** [comment]
Overriding the skip doesn't seem to work in the why I'd expect:

```
       matrix:
         external_net: ['external-net']
         globalnet: ['', 'globalnet']
+        skip: ['']
```

https://github.com/dfarrell07/submariner/pull/47 

```
2022-02-03T22:30:09.6497522Z ##[group]Run submariner-io/shipyard/gh-actions/e2e@devel
2022-02-03T22:30:09.6497858Z with:
2022-02-03T22:30:09.6498163Z   using: external-net globalnet
2022-02-03T22:30:09.6498442Z   k8s_version: 1.23
2022-02-03T22:30:09.6498786Z   working-directory: .
2022-02-03T22:30:09.6499138Z   target: e2e
2022-02-03T22:30:09.6499375Z   focus: \[.*\]
2022-02-03T22:30:09.6499695Z   skip: \[external-dataplane\]
2022-02-03T22:30:09.6499977Z   testdir: test/e2e
2022-02-03T22:30:09.6500273Z ##[endgroup]
```

**Comment by dfarrell07** [comment]
This should get the external network tests running: https://github.com/submariner-io/submariner/pull/1686

File: test/external/dataplane/gn_connectivity.go [hunk 15]
```diff
@@ -318,40 +338,43 @@ func testGlobalNetExternalConnectivity(p testParams, g globalnetTestParams) {
...
-		default:
-			fallthrough
-		case tcp.PodIP, tcp.ServiceIP:
-			framework.Failf("Unsupported ToEndpointType %v was passed", p.ToEndpointType)
-		case tcp.GlobalServiceIP:
-			By(fmt.Sprintf("Verifying that external app received request from one of podGlobalIPs %v", podGlobalIPs))
-			matchRegexp := MatchRegexp("%s .*GET /%s%s .*", podGlobalIPs[0], p.Framework.Namespace, clusterName)
-			for i := 1; i < len(podGlobalIPs); i++ {
-				matchRegexp = Or(matchRegexp, MatchRegexp("%s .*GET /%s%s .*", podGlobalIPs[i], p.Framework.Namespace, clusterName))
+	// TODO: Need to fix
```
Comment (mkimuram) [inline test/external/dataplane/gn_connectivity.go]: @sridhargaddam 

> When I'm trying to run these test-cases on my laptop, I'm getting the following errors (for all the external tests). Is it because of any docker version mismatch, any idea?

I've observed the same error here in my environment. So, I temporary disabled sending packets from cluster to external. I'll investigate more to enable it again.

File: test/external/dataplane/connectivity.go (hunk 8 above)
Comment (mkimuram) [reply test/external/dataplane/connectivity.go]: As I added comments in the code, this behavior won't be our current interest. So, we can go back again if we need this case to work well.
```
Note that the current expected use cases are:
    (a) From ext-app to non-gateway-cluster
    (b) From non-gateway-cluster to ext-app"
```
(Currently, we don't expect any pods on gateway-cluster to connect to ext-app, and it is just to be used as a gateway).

File: test/external/dataplane/connectivity.go (hunk 10 above)
Comment (mkimuram) [reply test/external/dataplane/connectivity.go]: Changed to `gatewayCluster`, but we may change the name again.

File: test/external/dataplane/gn_connectivity.go (hunk 12 above)
Comment (mkimuram) [reply test/external/dataplane/gn_connectivity.go:56]: > When you say S(*3) -> Pod w/ hostnetwork is not reachable when pod isn't on a gateway node -> Does it mean the ext-app is trying to connect to a Pod w/ hostnetwork?

Yes. 

> If it is so, then such use-case may not work even between regular clusters connected via submariner because the hostnetwork pod will have an IPAddress from the underlay network which is not advertised as local subnet to the remote clusters.

I'm not really sure how it works now, but it seems that it works only when the pod is on gateway node. I agree that we don't need to support hostnetwork (either on gateway node or non-gateway node).

File: test/external/dataplane/connectivity.go (hunk 13 above)
Comment (mkimuram) [reply test/external/dataplane/connectivity.go:52]: Any suggestion? I agree to change the name, but I failed to find a good wording.

File: test/external/dataplane/connectivity.go (hunk 14 above)
Comment (mkimuram) [reply test/external/dataplane/connectivity.go]: I agree that nesting to many `When` makes the log hard to read, but it would be easier to maintain the code. I meant to write the codes in the same way to [tcp_gn_pod_connectivity.go](https://github.com/submariner-io/submariner/blob/devel/test/e2e/dataplane/tcp_gn_pod_connectivity.go#L65). I've added `PWhen` to show that there are another pending cases. If it is better to remove pending cases and reduce the nest, please let me know (I don't have strong objection to it).

File: test/external/dataplane/gn_connectivity.go (hunk 11 above)
Comment (mkimuram) [reply test/external/dataplane/gn_connectivity.go:66]: This also only happens with the cases that we are not currently interested in.

File: test/external/dataplane/gn_connectivity.go (hunk 15 above)
Comment (mkimuram) [reply test/external/dataplane/gn_connectivity.go]: @sridhargaddam

I found that it is due to the change in removing KP dependency. As the implementation creates internal service, there is no endpoints for the internal service (Sharing the labels between original service and internal service makes the endpoints for service __with__ selector in sync, but there is no mechanism to sync endpoints for service __without__ selector by default).

I've confirmed that it works well, if endpoints for the internal service is created manually. So, I will consider creating a PR to sync from endpoints for original to one for internal.

<details>

- When service without selector is created,
```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster1
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: test-vm
spec:
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
EOF

cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-vm
subsets:
  - addresses:
      - ip: 172.117.0.4
    ports:
      - port: 80
EOF

subctl export service -n default test-vm
```

- Endpoints for original service exist, but endpoints for internal service don't exist. So, there will be no rule to forward for the endpoints.
```
# kubectl get svc,endpoints
NAME                                                  TYPE        CLUSTER-IP      EXTERNAL-IP     PORT(S)   AGE
service/kubernetes                                    ClusterIP   100.0.0.1       <none>          443/TCP   57m
service/submariner-4cl54k5nvyvsg3p65n5la4lbtg5scllc   ClusterIP   100.0.93.62     242.254.1.253   80/TCP    43m
service/test-vm                                       ClusterIP   100.0.120.156   <none>          80/TCP    43m

NAME                   ENDPOINTS         AGE
endpoints/kubernetes   172.18.0.8:6443   57m
endpoints/test-vm      172.117.0.4:80    43m
```

- So, if it is created manually, 
```
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: submariner-4cl54k5nvyvsg3p65n5la4lbtg5scllc
subsets:
  - addresses:
      - ip: 172.117.0.4
    ports:
      - port: 80
EOF
```

- Then, it works.
```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster2
kubectl -n default run tmp-shell --rm -i --tty --image quay.io/submariner/nettest -- bash
curl -m 10 242.254.1.253
```

</details>

Comment (mkimuram) [reply test/external/dataplane/gn_connectivity.go]: @sridhargaddam @nerdalert @anfredette

I've opened an issue(https://github.com/submariner-io/submariner/issues/1694) and a PR(https://github.com/submariner-io/submariner/pull/1695) for this.

**Comment by dfarrell07** [comment]
We're no longer skipping the external-network tests explicitly, but it seems they are still being skipped in 1 job because Globalnet isn't enabled.

```
2022-02-16T21:34:04.1009470Z === RUN   TestE2E
2022-02-16T21:34:04.1244458Z Running Suite: Submariner E2E suite
2022-02-16T21:34:04.1245223Z ===================================
2022-02-16T21:34:04.1245975Z Random Seed: [1m1645047244[0m - Will randomize all specs
2022-02-16T21:34:04.1246496Z Will run [1m19[0m of [1m26[0m specs
2022-02-16T21:34:04.1246666Z 
2022-02-16T21:34:04.1254181Z [1mSTEP[0m: Creating kubernetes clients
2022-02-16T21:34:04.1770288Z [1mSTEP[0m: Creating submariner clients
2022-02-16T21:34:04.1867128Z [0m[external-dataplane-globalnet] Connectivity[0m [90mwhen connected from a non-gateway-cluster[0m [0mwhen a pod matching an egress IP pod selector connects via TCP to/from the globalIP of a remote service[0m [90mwhen the pod is not on a gateway[0m 
2022-02-16T21:34:04.1992149Z   [1mshould be able to connect from/to an external app to/from a pod in a cluster[0m
2022-02-16T21:34:04.1992836Z   [37m/go/src/github.com/submariner-io/submariner/test/external/dataplane/gn_connectivity.go:83[0m
2022-02-16T21:34:04.2115246Z [1mSTEP[0m: Creating namespace objects with basename "ext-dataplane-gn"
2022-02-16T21:34:04.2115962Z [1mSTEP[0m: Generated namespace "e2e-tests-ext-dataplane-gn-jlh47" in cluster "cluster2" to execute the tests in
2022-02-16T21:34:04.2116892Z [1mSTEP[0m: Creating namespace "e2e-tests-ext-dataplane-gn-jlh47" in cluster "cluster1"
2022-02-16T21:34:04.2117582Z Feb 16 21:34:04.205: INFO: Globalnet is not enabled, skipping the test...
```

https://pipelines.actions.githubusercontent.com/XY3uESDBHEDMVySv1XHhnjBBj2HTkVMErmGSbcdh0GWkFVF5nH/_apis/pipelines/1/runs/18504/signedlogcontent/11?urlExpires=2022-02-16T21%3A42%3A57.3183904Z&urlSigningMethod=HMACV1&urlSignature=Kf9fWHhA8yYd%2BFWfD4cV8qsTnADZwt0n%2Bu5X2m29nTM%3D

I think we should just remove the external-net and no-globalnet part of the test matrix (I'll send a PR), only run with globalnet.

**Comment by dfarrell07** [comment]
> I think we should just remove the external-net and no-globalnet part of the test matrix (I'll send a PR), only run with globalnet.

https://github.com/submariner-io/submariner/pull/1697

**Comment by mkimuram** [comment]
@dfarrell07 

> We're no longer skipping the external-network tests explicitly, but it seems they are still being skipped in 1 job because Globalnet isn't enabled.

Thank you for checking it. Actually, it seems working as expected.

The skipped tests in  "(external-net)" are the ones for with globalnet and 
the skipped tests in  "(external-net, globalnet)" are the ones for without globalnet (Actually, if no one becomes interested in the "external-net without globalnet" use case, it makes sense to remove it from CI, but I'm currently not sure).

- External Network E2E (external-net)

```
        Feb 16 21:35:08.263: Globalnet is not enabled, skipping the test...

<snip>

Ran 6 of 26 Specs in 71.695 seconds
2775
SUCCESS! -- 6 Passed | 0 Failed | 7 Pending | 13 Skipped
```

- External Network E2E (external-net, globalnet)

```
        Feb 16 21:39:23.804: Globalnet is enabled, skipping the test...

<snip>

Ran 13 of 26 Specs in 269.618 seconds
2927
SUCCESS! -- 13 Passed | 0 Failed | 7 Pending | 6 Skipped
```

**Comment by anfredette** [comment]
> The skipped tests in "(external-net)" are the ones for with globalnet and the skipped tests in "(external-net, globalnet)" are the ones for without globalnet (Actually, if no one becomes interested in the "external-net without globalnet" use case, it makes sense to remove it from CI, but I'm currently not sure).

It seems less likely that people will use external networks without also using Globalnet because these are yet another set of networks that people would need to make sure don't overlap with any other networks at any other site.  However, if there's no overlap, it should just work, so I think I'd be inclined to leave the tests in.

**Comment by dfarrell07** [comment]
@anfredette @mkimuram Okay that makes sense, thanks for clarifying! Closing my PR.

**Comment by mkimuram** [comment]
Created another branch https://github.com/mkimuram/submariner/commits/issue/1581v2 that is rebased on https://github.com/submariner-io/submariner/pull/1695 . I will update this PR based on it, once the PR is merged.

**Comment by nyechiel** [comment]
> Created another branch https://github.com/mkimuram/submariner/commits/issue/1581v2 that is rebased on #1695 . I will update this PR based on it, once the PR is merged.

@mkimuram https://github.com/submariner-io/submariner/pull/1695 is now merged :)

**Comment by mkimuram** [comment]
@nyechiel Thank you for letting me know.

Rebased. PTAL @sridhargaddam @dfarrell07 @tpantelis @anfredette

**Comment by sridhargaddam** [comment]
The tests are now passing fine on my env:
```

Ran 13 of 26 Specs in 283.130 seconds
SUCCESS! -- 13 Passed | 0 Failed | 7 Pending | 6 Skipped
--- PASS: TestE2E (283.15s)
PASS
ok      github.com/submariner-io/submariner/test/external       283.160s
[16:05:23.187] [external]$ print_clusters_message
[16:05:23.190] [external]$ print_clusters_message
[16:05:23.192] [external]$ cat <<EOM
Your virtual cluster(s) are deployed and working properly and can be accessed with:

export KUBECONFIG=\$(find \$(git rev-parse --show-toplevel)/output/kubeconfigs/ -type f -printf %p:)

$ kubectl config use-context cluster1 # or cluster2, cluster3..

To clean everthing up, just run: make clean-clusters
EOM
Your virtual cluster(s) are deployed and working properly and can be accessed with:

export KUBECONFIG=$(find $(git rev-parse --show-toplevel)/output/kubeconfigs/ -type f -printf %p:)

$ kubectl config use-context cluster1 # or cluster2, cluster3..

To clean everthing up, just run: make clean-clusters
Successfully remade target file 'e2e'.
```

**Review by sridhargaddam — APPROVED** [review]

**Comment by sridhargaddam** [comment]
@mkimuram Thanks for the PR.

**Comment by sridhargaddam** [comment]
@anfredette @nerdalert can you please take a look at this PR?

**Review by nerdalert — COMMENTED** [review]
Gave it a spin and the tests looked good in my environment as well, ty!

```shell
Ran 13 of 26 Specs in 238.421 seconds
SUCCESS! -- 13 Passed | 0 Failed | 7 Pending | 6 Skipped
--- PASS: TestE2E (238.44s)
PASS
ok  	github.com/submariner-io/submariner/test/external	238.458s
[15:08:46.675] [external]$ print_clusters_message
[15:08:46.677] [external]$ print_clusters_message
[15:08:46.680] [external]$ cat <<EOM
Your virtual cluster(s) are deployed and working properly and can be accessed with:

export KUBECONFIG=\$(find \$(git rev-parse --show-toplevel)/output/kubeconfigs/ -type f -printf %p:)

$ kubectl config use-context cluster1 # or cluster2, cluster3..

To clean everthing up, just run: make clean-clusters
EOM
Your virtual cluster(s) are deployed and working properly and can be accessed with:

export KUBECONFIG=$(find $(git rev-parse --show-toplevel)/output/kubeconfigs/ -type f -printf %p:)

$ kubectl config use-context cluster1 # or cluster2, cluster3..

To clean everthing up, just run: make clean-clusters
Successfully remade target file 'e2e'.

```

**Comment by dfarrell07** [comment]
@nerdalert if this works for you, please vote +1 so we can merge it :)

**Review by nerdalert — APPROVED** [review]
LGTM, ty @mkimuram!

**Comment by mkimuram** [comment]
Thank you. Rebased.
